### ISSUE TITLE
Publishing doesn't trigger onBeforePublish or onAfterPublish in extensions

### ISSUE THREAD

**Issue by GuySartorelli** [dnadesign/silverstripe-elemental#779]
**Version: SS 4.4.3, CWP 2.4.0, and elemental 4.3.0**
Publishing an elemental block doesn't trigger the onBeforePublish or onAfterPublish functions for extension classes which are extending DataObject or BaseElement.
I initially thought this might be a problem with Versioned, but other Versioned DataObjects don't exhibit the problem behaviour.

**Steps to reproduce**
- Create a DataExtension which implements public functions onBeforePublish and onAfterPublish
- Add this extension to SilverStripe\ORM\DataObject or DNADesign\Elemental\Models\BaseElement
- dev/build and flush.
- Publish an elemental block (or the whole page, which will also publish the block)

**Expectation:** The onBeforePublish and onAfterPublish functions on the extension class will be called.
**Result:** Those functions are not called.

Note that while looking into this I have seen onBeforeWrite, onAfterWrite, onBeforeUnpublish, and onAfterUnpublish all trigger on the extension. It is only the publish hooks that I don't seem to be called.

# Pull requests
* [ ] https://github.com/dnadesign/silverstripe-elemental/pull/792

**Comment by maxime-rainville** [comment]
Confirmed the issue. Looking at the actual GraphQL request that publishes the blocks, it looks like where doing a copyToStage instead of doing a straight publish, which would explain why we're not triggering a the onBeforePublish/onAfterPublish.

This sounds like a really weird decision. It's not clear to me why we would choose to use a low level method like copyToStage instead of a plain publish.

```json
{   
   "operationName":"PublishBlock",
   "variables":{
      "blockId":"1",
      "fromStage":"DRAFT",
      "toStage":"LIVE",
      "fromVersion":12
   
   },
   "query":"
   mutation PublishBlock(
      $blockId: ID!,
      $fromStage: VersionedStage!,
      $toStage: VersionedStage!,
      $fromVersion: Int!
   ) {
      copyBlockToStage(
         Input: {ID: $blockId, FromVersion: $fromVersion, FromStage: $fromStage, ToStage: $toStage}
      ) {
         ID
         __typename
      }
   }
"
}

**Comment by maxime-rainville** [comment]
I'm marking this as impact high because it forecloses a very common and high value use case for developers.

**Comment by barrykeenan** [comment]
I have a similar issue adding my own DataExtension to a Page, that expects onBeforePublish and onAfterPublish hooks to be called. 

When just a standard Page, if the Content field is edited and the Page is published: onBeforePublish and onAfterPublish hooks are called correctly.

But when the Page has an ElementalPageExtension applied, if any child Elements are edited and the Page published: onBeforePublish and onAfterPublish hooks are not called.

Can you suggest a workaround to trigger these, perhaps by extending ElementalPageExtension or something?

Many thanks,
Barry

**Comment by GuySartorelli** [comment]
Child elements will not trigger the publish hooks of an extension which was applied to the SiteTree - you'll need an extension for DataObject, BaseElement, or the specific class(es) you're interested in. It's at _that_ stage you hit the issue described above.

If it doesn't specifically matter that you're _publishing_ (i.e. the changes affected by your onBeforePublish or onAfterPublish method are allowed to occur on drafts) you can use onBeforeWrite and onAfterWrite which do function as expected.
Alternatively you could try to add into your business practice (at least until this is resolved) that the parent page is also published when a child is published - assuming there are no pending changes on that page which should not yet be published.

If you really need this asap you could fork this repo and apply a fix (and also submit it as a pull request here ;P)

**Comment by barrykeenan** [comment]
Hiya @GuySartorelli - ah thanks for that! Turns out I did need an extension on BaseElement to detect when these are being published. So now I am actually hitting the problem above. Cheers for the help tho 😄

**Comment by matt-in-a-hat** [comment]
Any ideas on how to get around this?

I could hook into `onAfterPublish` on BaseElement however this will happen for every element that is being published. In my use-case I want to update a search index when a page is published, but the page `onAfterPublish` isn't triggered when only an elemental block is changed, and `onAfterPublish` on Elements would mean I'd be updating the search index multiple times per save (unless I debounce it somehow? 🤔)

**Comment by maxime-rainville** [comment]
I added a PR that switch to using the publishBlock mutation instead of the copyBlockToStage mutation. https://github.com/dnadesign/silverstripe-elemental/pull/792

This seems to work fine at first glance, but I didn't put much thought into this. Maybe there's an obvious problem why this can't work that eludes why puny intellect.

**Comment by maxime-rainville** [comment]
I dug up why we're using copyToStage instead of a plain publish. Guy says that's because when the publish feature was being implemented, they had to support a version of Silverstripe that didn't have the versioned publish mutation yet. https://github.com/dnadesign/silverstripe-elemental/pull/360#discussion_r412671953

The [publish mutation got shipped with versioned 1.1](https://github.com/silverstripe/silverstripe-versioned/commit/5892acfb958e4d5eade675689e7be94d7aa0c282#diff-18982da4e3da28e02b3e0c6820ff95a6). Elemental currently requires versioned-admin 1.2, which requires versioned 1.3.

So long story short, it's totally kosher for us to switch to `publishBlock`.

**Comment by ScopeyNZ** [comment]
Yeah it was a long time ago and the GraphQL library was going through some pretty rapid changes. I can't remember what base version of SilverStripe we were targeting with Elemental 4.

I can't think of any good reason that we use "copyToStage" over "publish" and the commits/PRs that made the change don't seem to offer any reason.

**Comment by GuySartorelli** [comment]
Now that #792 has been merged, should this be closed?

**Comment by maxime-rainville** [comment]
Yep. Should be available in the next release.

**Comment by thats4shaw** [comment]
@matt-in-a-hat did you come up with a solution for your dilemma?  I've just struck the same problem.

**Comment by matt-in-a-hat** [comment]
@thats4shaw I ended up adding a save hook to BaseElement (below) and debouncing the index calls using queuedjobs (which was already supported by the Aloglia index addon)

```php
public function onAfterPublish()
{
    // Currently SiteTree::onAfterPublish isn't called when only
    // an elemental block has changed (and no direct page fields).
    // This means AlgoliaObjectExtension doesn't index elemental changes.

    // This hook gets called when any elemental block is published.
    // However that could be multiple times on any 1 page publish.
    // Using the option AlgoliaObjectExtension::use_queued_indexing
    // which means that on publish we can call indexInAlgolia many
    // times, yet it will only queue up 1 job for this owner page,
    // essentially debouncing the all the publish calls from blocks.

    $this->owner->getPage()->indexInAlgolia();
}
```

Not ideal and there certainly could be other use cases where debouncing is much harder or not possible.
